### PR TITLE
Jax pipeline

### PR DESCRIPTION
**Pull request by b-biswas**
TODO: 

- [x] `_get_subobs_jax` is not jitable: need to drop checking boundary conditions
- [x] Return proper detection and boundary flags
- [x] vmap over fitting moments to each galaxy 
- [x] Write tests for `jax_single_band_deep_field_metadetect` when using the peak finder
- [x] Set appropriate peakfinder noise threshold

### CODE REVIEW

File: deep_field_metadetect/gaussmom/gaussmom.py [hunk 1]
```diff
@@ -483,10 +477,11 @@ def get_ratio_error(a, b, var_a, var_b, cov_ab):
...
 
 
-@dataclass
+@jax.tree_util.register_pytree_node_class
 class GaussMom:
-    fwhm: float
-    with_higher_order: bool = False
+    def __init__(self, fwhm: float, with_higher_order: bool = False):
+        self.fwhm = fwhm
+        self.with_higher_order = with_higher_order
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom.py]: Let's leave this as an option to the method and not an option to the class.

File: deep_field_metadetect/gaussmom/gaussmom_core.py [hunk 2]
```diff
@@ -154,3 +156,22 @@ def obs_to_gaussmom_obs(obs: ngmix.Observation) -> GaussMomObs:
...
+    dx = x - (obs.wcs.origin.x - 1)
+    dy = y - (obs.wcs.origin.y - 1)
+    u = obs.wcs.dudx * dx + obs.wcs.dudy * dy
+    v = obs.wcs.dvdx * dx + obs.wcs.dvdy * dy
+
+    return GaussMomObs(
+        u,
+        v,
+        obs.image,
+        obs.wcs.dudx * obs.wcs.dvdy - obs.wcs.dudy * obs.wcs.dvdx,
```
Comment (beckermr) [inline deep_field_metadetect/gaussmom/gaussmom_core.py]: We need an absolute value here. I may have missed this above. Also for galsim objects, there is the `.pixelArea()` method.

File: deep_field_metadetect/metadetect.py [hunk 3]
```diff
@@ -78,6 +80,11 @@ def single_band_deep_field_metadetect(
         To fix max and min values of FFT size.
         Defaults to None which lets Galsim determine the values.
         Used mainly to test against Galsim.
+    return_debug_info: bool
+        return detections and mcal_res for debugging
+    debug_verbose: bool
```
Comment (beckermr) [inline deep_field_metadetect/metadetect.py]: I am fine with using print statements, but we may want to move to logging eventually.

File: deep_field_metadetect/metadetect.py [hunk 4]
```diff
@@ -113,13 +120,16 @@ def single_band_deep_field_metadetect(
 
     psf_res = fit_gauss_mom_obs(mcal_res["noshear"].psf)
     dfmdet_res = []
+    detections = []
     for shear in shears:
         obs = mcal_res[shear]
         detres = run_detection_sep(obs, nodet_flags=nodet_flags)
-
+        if debug_verbose:
+            print("num detections : " + str(len(detres["catalog"]["x"])))
```
Comment (beckermr) [inline deep_field_metadetect/metadetect.py]:
```suggestion
            print("num detections:", len(detres["catalog"]["x"]))
```

File: deep_field_metadetect/metadetect.py [hunk 5]
```diff
@@ -144,6 +154,8 @@ def single_band_deep_field_metadetect(
                 (ind + 1, obj["x"], obj["y"], shear, bmask_flags[ind], mfrac_vals[ind])
                 + tuple(fres[0])
             )
+            if debug_verbose:
+                print(str(ind) + "non-jax tuple:" + str(tuple(fres[0])))
```
Comment (beckermr) [inline deep_field_metadetect/metadetect.py]:
```suggestion
                print(str(ind) + " non-jax tuple:", tuple(fres[0]))
```

File: deep_field_metadetect/metadetect.py [hunk 6]
```diff
@@ -154,6 +166,16 @@ def single_band_deep_field_metadetect(
...
+            return (
+                (np.array(dfmdet_res, dtype=total_dtype), kinfo),
+                mcal_res,
+                detections,
+            )
+
+        return np.array(dfmdet_res, dtype=total_dtype), mcal_res, detections
+
     if return_k_info:
         return (np.array(dfmdet_res, dtype=total_dtype), kinfo)
```
Comment (beckermr) [inline deep_field_metadetect/metadetect.py]: Did we discuss returning a dictionary here as the second item and adding keys to that as needed? I don't recall and you may have suggested this earlier and I may have suggested something else. 

Regardless, as we add more return options here, it might make sense to move to that format to keep the number of return arguments constant.

File: deep_field_metadetect/jaxify/tests/test_jax_metadetect.py [hunk 7]
```diff
@@ -417,7 +426,7 @@ def test_metadetect_single_band_deep_field_metadetect_mfrac_deep():
 
 @pytest.mark.parametrize("deep_psf_ratio", [0.8, 1, 1.1])
 def test_metadetect_single_band_deep_field_metadetect(deep_psf_ratio):
-    nsims = 100
+    nsims = 25
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/tests/test_jax_metadetect.py]: Maybe we can put this back?

File: deep_field_metadetect/jaxify/jax_metacal.py [hunk 8]
```diff
@@ -919,14 +932,4 @@ def jax_metacal_wide_and_deep_psf_matched(
         fft_size=fft_size,
     )
 
-    if return_k_info:
-        mcal_res, kinfo = mcal_res
-
-    for k in shears:
-        mcal_res[k] = dfmd_obs_to_ngmix_obs(mcal_res[k])
-        mcal_res[k].psf.galsim_obj = reconv_psf
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_metacal.py:927]: What motivates the change in return type here? I guess this function is now fed into other jax code?

File: deep_field_metadetect/jaxify/observation.py [hunk 9]
```diff
@@ -350,3 +351,282 @@ def dfmd_obs_to_ngmix_obs(dfmd_obs: DFMdetObservation) -> Observation:
...
+
+# Constants for padding
+BMASK_EDGE = 2**30
+DEFAULT_IMAGE_VALUES = {
+    "image": 0.0,
+    "weight": 0.0,
+    "bmask": BMASK_EDGE,
+    "noise": 0.0,
+    "mfrac": 0.0,
+}
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/observation.py]: I think these constants may be define elsewhere, especially BMASK_EDGE. We should unify wherever possible.

File: deep_field_metadetect/jaxify/jax_dfmd_defaults.py [hunk 10]
```diff
@@ -6,3 +6,5 @@
 
 DEFAULT_RECONV_DK = compute_dk(pixel_scale=0.2, image_size=53)
 DEFAULT_KIM_SIZE = compute_kim_size(image_size=DEFAULT_NXY_PSF)
+
+MAX_OBJECTS = 20  # Max number of objects in a field
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_dfmd_defaults.py]: This number is surprisingly small. Shouldn't it be set adaptively as the number of pixels or a fraction of that?

File: deep_field_metadetect/jaxify/jax_detection.py [hunk 11]
```diff
@@ -69,6 +77,9 @@ def peak_finder(
     positions : jnp.ndarray
         Array of peak coordinates (y, x) of shape (max_objects, 2)
         Invalid entries filled with (-1, -1)
+    det_flag : jnp.ndarray
+        Integer array of shape (max_objects,) where:
+        0 = actual detection, 1 = fill value
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_detection.py]: This is totally opposite of my normal expectations.

File: deep_field_metadetect/jaxify/jax_detection.py [hunk 12]
```diff
@@ -78,7 +89,11 @@ def peak_finder(
 
     positions = jnp.argwhere(local_max_mask, size=max_objects, fill_value=(-1, -1))
 
-    return positions
+    # Create detection flag: 0 for actual detections, 1 for fill values
+    # Fill values have (-1, -1) coordinates
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_detection.py]: Filling with -1, -1 seems problematic. Why not use nans?

File: deep_field_metadetect/jaxify/jax_detection.py [hunk 13]
```diff
@@ -428,3 +468,272 @@ def scan_fn(markers_current, i_peak):
...
+        mfrac=sub_mfrac,
+        wcs=new_wcs,
+        psf=obs.psf,
+        meta=obs.meta,
+        store_pixels=obs.store_pixels,
+        ignore_zero_weight=obs.ignore_zero_weight,
+    )
+
+
+def jax_generate_subobs_for_detections(
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_detection.py]: Do all of the generation functions here get used? Can we only use the vmapped one and get rid of the rest?

File: deep_field_metadetect/jaxify/jax_detection.py [hunk 14]
```diff
@@ -428,3 +468,272 @@ def scan_fn(markers_current, i_peak):
...
+        """Process one detection given (x, y) coordinates."""
+        x, y = xy
+        subobs = extract_single_detection(x, y)
+        return carry, subobs
+
+    xy_pairs = jnp.stack([xs, ys], axis=1)
+
+    # Use scan to iterate through all detections
+    # TODO: see memory vs speed trade off on using vmap
+    _, all_subobs = jax.lax.scan(scan_fn, None, xy_pairs)
```
Comment (beckermr) [inline deep_field_metadetect/jaxify/jax_detection.py]: You could use the unroll parameter here to allow folks to control memory vs speed here maybe?

**Review by beckermr — COMMENTED** [review]
Looking great! A few comments and questions.

File: deep_field_metadetect/jaxify/jax_detection.py (hunk 12 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/jax_detection.py]: The moments computation is done on images extracted at these locations. After padding, we can still extract stamps at (-1,-1) without hitting the boundary so we can jit the remaining section of the pipeline more easily at the cost of carrying around a detection_flag

File: deep_field_metadetect/jaxify/jax_detection.py (hunk 13 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/jax_detection.py]: This last function is outdated, I was using it when ObsList and MultiBandObsList classes were not defined. I am removing this.

File: deep_field_metadetect/jaxify/jax_metacal.py (hunk 8 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/jax_metacal.py:927]: Yes, it was to the jit the `jax_metacal_wide_and_deep_psf_matched` function. I did not save the reconv psf as `mcal_res[k].psf.galsim_obj` as of now because we have a custom PSF class. But I can modify the output of this function to return it.

File: deep_field_metadetect/jaxify/jax_detection.py (hunk 11 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/jax_detection.py]: I flipped the convention here along with the `refinement_flag`

File: deep_field_metadetect/metadetect.py (hunk 6 above)
Comment (b-biswas) [reply deep_field_metadetect/metadetect.py]: changed to using dicts

File: deep_field_metadetect/jaxify/observation.py (hunk 9 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/observation.py]: Now loading them from the non-JAX version. [They are defined in deep_field_metadetect.detect]

File: deep_field_metadetect/jaxify/jax_dfmd_defaults.py (hunk 10 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/jax_dfmd_defaults.py]: Done! now computed according to the number density and the field size

File: deep_field_metadetect/jaxify/tests/test_jax_metadetect.py (hunk 7 above)
Comment (b-biswas) [reply deep_field_metadetect/jaxify/tests/test_jax_metadetect.py]: Keeping it at 100 runs the git CI for too long. I have added this as a comment
